webrtc/webrtc.md

```diff
@@ -0,0 +1,121 @@
+# WebRTC
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+|-----------------|---------------|--------|-----------------|
+| 1A              | Working Draft | Active | r0, 2023-03-03  |
+
+Authors: [@mxinden]
+
+## Motivation
+
+libp2p transport protocol enabling two private nodes (e.g. two browsers) to establish a direct connection.
+
+Browser _A_ wants to connect to Browser node _B_ with the help of server node _R_.
+Both _A_ and _B_ can not listen for incoming connections due to running in a constrained environment (i.e. a browser) with its only transport capability being the W3C WebRTC `RTCPeerConnection` API and being behind a NAT and/or firewall.
+Note that _A_ and/or _B_ may as well be non-browser nodes behind NATs and/or firewalls.
+However, for two non-browser nodes using TCP or QUIC hole punching with [DCUtR] will be the more efficient way to establish a direct connection.
+
+On a historical note, this specification replaces the existing [libp2p WebRTC star](https://github.com/libp2p/js-libp2p-webrtc-star) and [libp2p WebRTC direct](https://github.com/libp2p/js-libp2p-webrtc-direct) protocols.
+
+## Connection Establishment
+
+1. _B_ advertises support for the WebRTC browser-to-browser protocol by appending `/webrtc` to its relayed multiaddr e.g. `/ip6/fe80::883:a581:fff1:833/udp/4001/quic/webtransport/certhash/<hash1>/p2p/<relay-peer-id>/p2p-circuit/webrtc/p2p/<b-peer-id>`.
+
+2. Upon discovery of _B_'s multiaddress, _A_ learns that _B_ supports the WebRTC transport through the (constrained) W3C API and knows how to establish a relayed connection to _B_ to run the `/webrtc-signaling` protocol on top.
+
+3. _A_ establish a relayed connection to _B_.
+   Note that further steps depend on the relayed connection to be authenticated, i.e. that data sent on the relayed connection can be trusted.
+
+4. _A_ (outbound side of relayed connection) creates an `RTCPeerConnection` provided by a W3C compliant WebRTC implementation (e.g. a browser).
+   See [STUN](#stun) section on what STUN servers to configure at creation time.
+   _A_ creates an SDP offer via `RTCPeerConnection.createOffer()`.
+   _A_ initiates the signaling protocol to _B_ via the relayed connection from (1), see [Signaling Protocol](#signaling-protocol) and sends the offer to _B_.
```

Comment:
Do we need to restrict who can initiate the WebRTC connection? It would be nice if on any relayed connection, any of those nodes could decide to try a WebRTC connection attempt.

Reply:
I am in favor of not restricting it.

Peers can also just try to speak the signaling protocol. We should keep in mind that advertising via `/webrtc` is just one way.

A node may not advertise via the listen address but still speak the signaling protocol.

Reply:
@marten-seemann @thomaseizinger how does one prevent both peers to start the process and thus two direct WebRTC connections being established?

I suggest we stick with the convention that _A_ initiates the signaling protocol, though require _A_ to also be able to accept an incoming signaling protocol stream in case we ever want to support the process from the other direction (_B_ -> _A_).

Reply:
Extended through 01f78a7. Let me know what you think.

Reply:
Extended through 01f78a7. Let me know what you think.

Reply:
> Extended through [01f78a7](https://github.com/libp2p/specs/commit/01f78a735765fc33ecc25f56287148ebf8aa91de). Let me know what you think.

Deduping is indeed an issue. I like this new text since we're still allowing any side to initiate the request, this might become useful later.